File: sentil-java/src/main/java/io/github/sedislab/sentil/FormulaBank.java
package io.github.sedislab.sentil;

import java.util.Arrays;
import java.util.List;
import java.util.Map;

/** A batch of named formulas evaluated together over one trace. */
public final class FormulaBank extends NativeResource {
    public FormulaBank() throws SentilException {
        super(NativeLib.formulaBankCreate(), NativeLib::formulaBankDestroy);
    }

    /** Add a formula string under an id. */
    public void add(String id, String formula) throws SentilException {
        NativeLib.formulaBankAdd(handle(), id, formula);
    }

    /** Add a formula under an id; the formula is borrowed, not consumed. */
    public void add(String id, Formula formula) throws SentilException {
        NativeLib.formulaBankAddFormula(handle(), id, formula.handle());
    }

    /** The ids in insertion order. */
    public List<String> ids() throws SentilException {
        return Arrays.asList(NativeLib.formulaBankIds(handle()));
    }

    /** The number of formulas. */
    public long size() {
        return NativeLib.formulaBankLen(handle());
    }

    /** Whether no formula is registered. */
    public boolean isEmpty() {
        return NativeLib.formulaBankIsEmpty(handle());
    }

    /** The robustness of every formula over the trace, keyed by id. */
    public Map<String, Double> robustness(Trace trace) throws SentilException {
        return NativeLib.formulaBankRobustness(handle(), trace.handle());
    }

    /** The dense-time robustness of every formula over the trace, keyed by id. */
    public Map<String, Double> robustnessDense(Trace trace) throws SentilException {
        return NativeLib.formulaBankRobustnessDense(handle(), trace.handle());
    }
}

File: sentil-matlab/+sentil/FormulaBank.m
classdef FormulaBank < handle
    %FORMULABANK A batch of named formulas evaluated together over one trace.

    properties (SetAccess = private, Hidden)
        Handle uint64 = uint64(0)
    end

    methods
        function obj = FormulaBank()
            obj.Handle = sentil_mex('formula_bank_create');
        end

        function delete(obj)
            if obj.Handle ~= 0
                sentil_mex('formula_bank_destroy', obj.Handle);
                obj.Handle = uint64(0);
            end
        end

        function add(obj, id, spec)
            %ADD Register a formula under an id, from a string or a borrowed sentil.Formula.
            obj.assertOpen();
            if isa(spec, 'sentil.Formula')
                sentil_mex('formula_bank_add_formula', obj.Handle, char(id), spec.Handle);
            else
                sentil_mex('formula_bank_add', obj.Handle, char(id), char(spec));
            end
        end

        function v = ids(obj)
            %IDS The registered ids in insertion order.
            obj.assertOpen();
            v = sentil_mex('formula_bank_ids', obj.Handle);
        end

        function n = length(obj)
            %LENGTH The number of formulas.
            obj.assertOpen();
            n = sentil_mex('formula_bank_len', obj.Handle);
        end

        function tf = is_empty(obj)
            %IS_EMPTY Whether no formula is registered.
            obj.assertOpen();
            tf = sentil_mex('formula_bank_is_empty', obj.Handle);
        end

        function m = robustness(obj, trace)
            %ROBUSTNESS The robustness of every formula over the trace, keyed by id.
            obj.assertOpen();
            m = obj.toMap(sentil_mex('formula_bank_robustness', obj.Handle, trace.Handle));
        end

        function m = robustness_dense(obj, trace)
            %ROBUSTNESS_DENSE The dense-time robustness of every formula, keyed by id.
            obj.assertOpen();
            m = obj.toMap(sentil_mex('formula_bank_robustness_dense', obj.Handle, trace.Handle));
        end
    end

    methods (Static, Access = private)
        function m = toMap(result)
            if isempty(result.ids)
                m = containers.Map('KeyType', 'char', 'ValueType', 'double');
            else
                m = containers.Map(result.ids, num2cell(result.values));
            end
        end
    end

    methods (Access = private)
        function assertOpen(obj)
            if obj.Handle == 0
                error('sentil:handle', 'this bank has been closed');
            end
        end
    end
end